src/postgraphile/http/subscriptions.ts

```diff
@@ -186,134 +191,262 @@ export async function enhanceHttpServerWithSubscriptions<
-        // eslint-disable-next-line require-atomic-updates
-        params.formatResponse = formatResponse;
-        const hookedParams = pluginHook
-          ? pluginHook('postgraphile:ws:onOperation', params, {
-              message,
-              params,
-              socket,
-              options,
-            })
-          : params;
-        const finalParams: typeof hookedParams & { query: DocumentNode } = {
-          ...hookedParams,
-          query:
-            typeof hookedParams.query !== 'string' ? hookedParams.query : parse(hookedParams.query),
-        };
-
-        // You are strongly encouraged to use
-        // `postgraphile:validationRules:static` if possible - you should
-        // only use this one if you need access to variables.
-        const moreValidationRules = pluginHook('postgraphile:validationRules', [], {
-          options,
-          req,
-          res,
-          variables: params.variables,
-          operationName: params.operationName,
-          meta,
-        });
-        if (moreValidationRules.length) {
-          const validationErrors: ReadonlyArray<GraphQLError> = validate(
-            params.schema,
-            finalParams.query,
-            moreValidationRules,
-          );
-          if (validationErrors.length) {
-            const error = new Error(
-              'Query validation failed: \n' + validationErrors.map(e => e.message).join('\n'),
+            let normalizedConnectionParams = {};
+            if (connectionParams) {
+              normalizedConnectionParams = lowerCaseKeys(connectionParams);
+              request['connectionParams'] = connectionParams;
+              request['normalizedConnectionParams'] = normalizedConnectionParams;
+            }
+
+            if (!request.headers.authorization && normalizedConnectionParams['authorization']) {
+              /*
+               * Enable JWT support through connectionParams.
+               *
+               * For other headers you'll need to do this yourself for security
+               * reasons (e.g. we don't want to allow overriding of Origin /
+               * Referer / etc)
+               */
+              request.headers.authorization = String(normalizedConnectionParams['authorization']);
+            }
+
+            socket['postgraphileHeaders'] = {
+              ...normalizedConnectionParams,
+              // The original headers must win (for security)
+              ...request.headers,
+            };
+          },
+          async onSubscribe(ctx, msg) {
+            const context = await getContext(ctx.socket, msg.id);
+
+            // Override schema (for --watch)
+            const schema = await getGraphQLSchema();
+
+            const { payload } = msg;
+            const args: ExecutionArgs = {
+              schema,
+              contextValue: context,
+              operationName: payload.operationName,
+              document: parse(payload.query),
```

Comment:
In the case of persisted operations, `payload.query` may not be present, it might be `payload.documentId` or something else for example. I'm not sure, but I think that calling `parse` here might throw an error if there is no `query`? We want effectively `document: (() => {try{return parse(payload.query)} catch { return null }})()`

Reply:
When using persisted queries, you wont be _just_ parsing the payload's query. The preparation should happen beforehand, like checking the `documentId` field and supplying your own document during assembly. Would you agree?

You can check the ["Server and client usage with persisted queries" recipe](https://github.com/enisdenjo/graphql-ws#recipes) on how I envisioned this.

Reply:
In that example you override it in the `onSubscribe` hook, which is what I'd like to do too. In the existing implementation we use the `postgraphile:ws:onOperation` hook for this, so I'd expect to achieve it through the `postgraphile:ws:onSubscribe` hook a couple lines below here - passing the args (sans valid query) and message (with `documentId` or whatever) and getting back in `hookedArgs` the valid args with new `query` property set.

Reply:
I expect it'd look very much like this: https://github.com/graphile/persisted-operations/blob/ee2e404ccc5e4f7f7ffe65215a97ac5e6e0fc08e/src/index.ts#L256-L263

Reply:
Didn't saw that there! I'll make sure to add support for persisted queries using the `onSubscribe` plugin hook.

Reply:
With https://github.com/graphile/postgraphile/pull/1338/commits/1074b013657302773054850a784578e3c2d8f4f5, you should be able to do something like:
```ts
"postgraphile:ws:onSubscribe" = (args, { message, options }) => {
  args.query = persistedOperationFromPayload(
    message.payload, // graphql-ws subscribe message payload
    options
  );
  return args;
};
```

Reply:
That looks reasonable. It prevents passing the query identifier via the `query` parameter but I think that'd be a bit of an undesirable thing to do anyway. Apollo Client uses the `extensions` property on the message, and Relay uses a `documentId` field, and I've seen it just done via an `id` field before, so I think we're fine on this.